cpp/src/arrow/c/bridge_test.cc

```diff
@@ -4746,4 +4750,531 @@ TEST_F(TestArrayStreamRoundtrip, ChunkedArrayRoundtripEmpty) {
+  auto orig_schema = arrow::schema({field("ints", int32())});
+  auto batches = MakeBatches(orig_schema, {arr1, arr2});
+  ASSERT_OK_AND_ASSIGN(
+      auto reader, RecordBatchReader::Make(batches, orig_schema, device->device_type()));
+
+  Roundtrip(std::move(reader), [&](const std::shared_ptr<RecordBatchReader>& reader) {
+    AssertSchemaEqual(*orig_schema, *reader->schema(), /*check_metadata=*/true);
+    AssertReaderNext(reader, *batches[0]);
+    AssertReaderNext(reader, *batches[1]);
+    AssertReaderEnd(reader);
+    AssertReaderEnd(reader);
+    AssertReaderClose(reader);
+  });
+}
+
+TEST_F(TestArrayDeviceStreamRoundtrip, CloseEarly) {
+  std::shared_ptr<Device> device = std::make_shared<MyDevice>(1);
+  auto mm = device->default_memory_manager();
+
+  ASSERT_OK_AND_ASSIGN(auto arr1,
+                       ToDevice(mm, *ArrayFromJSON(int32(), "[1, 2]")->data()));
+  ASSERT_EQ(device->device_type(), arr1->device_type());
+  ASSERT_OK_AND_ASSIGN(auto arr2,
+                       ToDevice(mm, *ArrayFromJSON(int32(), "[4, 5, null]")->data()));
+  ASSERT_EQ(device->device_type(), arr2->device_type());
+  auto orig_schema = arrow::schema({field("ints", int32())});
+  auto batches = MakeBatches(orig_schema, {arr1, arr2});
+  ASSERT_OK_AND_ASSIGN(
+      auto reader, RecordBatchReader::Make(batches, orig_schema, device->device_type()));
+
+  Roundtrip(std::move(reader), [&](const std::shared_ptr<RecordBatchReader>& reader) {
+    AssertReaderNext(reader, *batches[0]);
+    AssertReaderClose(reader);
+  });
+}
+
+TEST_F(TestArrayDeviceStreamRoundtrip, Errors) {
+  auto reader = std::make_shared<FailingRecordBatchReader>(
+      Status::Invalid("roundtrip error example"));
+
+  Roundtrip(std::move(reader), [&](const std::shared_ptr<RecordBatchReader>& reader) {
+    auto status = reader->Next().status();
+    ASSERT_RAISES(Invalid, status);
+    ASSERT_THAT(status.message(), ::testing::HasSubstr("roundtrip error example"));
+  });
+}
+
+TEST_F(TestArrayDeviceStreamRoundtrip, SchemaError) {
+  struct StreamState {
+    bool released = false;
+
+    static const char* GetLastError(struct ArrowDeviceArrayStream* stream) {
+      return "Expected error";
+    }
+
+    static int GetSchema(struct ArrowDeviceArrayStream* stream,
+                         struct ArrowSchema* schema) {
+      return EIO;
+    }
+
+    static int GetNext(struct ArrowDeviceArrayStream* stream,
+                       struct ArrowDeviceArray* array) {
+      return EINVAL;
+    }
+
+    static void Release(struct ArrowDeviceArrayStream* stream) {
+      reinterpret_cast<StreamState*>(stream->private_data)->released = true;
+      std::memset(stream, 0, sizeof(*stream));
+    }
+  } state;
+  struct ArrowDeviceArrayStream stream = {};
+  stream.get_last_error = &StreamState::GetLastError;
+  stream.get_schema = &StreamState::GetSchema;
+  stream.get_next = &StreamState::GetNext;
+  stream.release = &StreamState::Release;
+  stream.private_data = &state;
```

Comment:
Empty lambdas are convertible to function pointers:
```suggestion
  struct ArrowDeviceArrayStream stream = {};
  stream.get_last_error = [](struct ArrowDeviceArrayStream* stream) {
    return "Expected error";
  };
  stream.get_schema = [](struct ArrowDeviceArrayStream* stream,
                         struct ArrowSchema* schema) {
    return EIO;
  };
  stream.get_next = [](struct ArrowDeviceArrayStream* stream,
                       struct ArrowDeviceArray* array) {
    return EINVAL;
  };
  stream.release = [](struct ArrowDeviceArrayStream* stream) {
    *static_cast<bool*>(stream->private_data) = true;
    std::memset(stream, 0, sizeof(*stream));
  };
  bool released = false;
  stream.private_data = &released;
```